### PR TITLE
Metrics instrumentation starlette

### LINKED ISSUE #1146: Metrics instrumentation starlette
HTTP Server semconv https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#http-server

### PR DESCRIPTION
**Pull request by rahulmukherjee68**
# Description

Fixes #1146 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] tox -e test-instrumentation-starlette

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-starlette/tests/test_starlette_instrumentation.py [hunk 1]
```diff
@@ -100,6 +116,71 @@ def test_starlette_excluded_urls(self):
...
+                        if isinstance(point, NumberDataPoint):
+                            number_data_point_seen = True
+                        for attr in point.attributes:
+                            self.assertIn(
+                                attr, _recommended_attrs[metric.name]
+                            )
+        self.assertTrue(number_data_point_seen and histogram_data_point_seen)
+
+    def test_basic_post_request_metric_success(self):
+        start = default_timer()
```
Comment (sanketmehta28) [inline instrumentation/opentelemetry-instrumentation-starlette/tests/test_starlette_instrumentation.py:148]: add assert statement for metric attributes

**Review by sanketmehta28 — CHANGES_REQUESTED** [review]

File: instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py [hunk 2]
```diff
@@ -186,20 +205,46 @@ def _instrument(self, **kwargs):
         _InstrumentedStarlette._client_response_hook = kwargs.get(
             "client_response_hook"
         )
+        _InstrumentedStarlette._meter_provider = kwargs.get("_meter_provider")
+
         applications.Starlette = _InstrumentedStarlette
 
+        for instance in _InstrumentedStarlette._instrumented_starlette_apps:
+            self.instrument_app(
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py:213]: @srikanthccv 
Do we want to instrument all currently existing apps if we call `instrument()` or only apps created AFTER `instrument()` was called?

File: instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py [hunk 3]
```diff
@@ -186,20 +205,46 @@ def _instrument(self, **kwargs):
...
+                instance,
+                server_request_hook=_InstrumentedStarlette._server_request_hook,
+                client_request_hook=_InstrumentedStarlette._client_request_hook,
+                client_response_hook=_InstrumentedStarlette._client_response_hook,
+                tracer_provider=_InstrumentedStarlette._tracer_provider,
+            )
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py:219]: Pass in `meter_provider`?

File: instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py [hunk 4]
```diff
@@ -186,20 +205,46 @@ def _instrument(self, **kwargs):
...
+
     def _uninstrument(self, **kwargs):
+
+        """uninstrumenting all created apps by user"""
+        for instance in _InstrumentedStarlette._instrumented_starlette_apps:
+            self.uninstrument_app(instance)
+
+        _InstrumentedStarlette._instrumented_starlette_apps = {
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py:227]: `_InstrumentedStarlette._instrumented_starlette_apps .clear()`

File: instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py (hunk 2 above)
Comment (srikanthccv) [reply instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py:213]: I would expect it to instrument the apps created after the instrument is invoked.

File: instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py (hunk 4 above)
Comment (rahulmukherjee68) [reply instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py:227]: sure will add this

File: instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py (hunk 3 above)
Comment (rahulmukherjee68) [reply instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py:219]: sure will add this too. Forgot to add it

File: instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py (hunk 2 above)
Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py]: @rahulmukherjee68 
Might want to remove this behavior.

Comment (rahulmukherjee68) [reply instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py]: Yeah @lzchen i have removed it in my latest commits

File: instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py [hunk 5]
```diff
@@ -168,9 +172,24 @@ def instrument_app(
...
+            # adding apps to set for uninstrumenting
+            if app not in _InstrumentedStarlette._instrumented_starlette_apps:
+                _InstrumentedStarlette._instrumented_starlette_apps.add(app)
+
+    @staticmethod
+    def uninstrument_app(app: applications.Starlette):
+        app.user_middleware = [
+            x
+            for x in app.user_middleware
+            if x.cls is not OpenTelemetryMiddleware
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py:188]: Should we do `is not instanceof()` here?

**Review by lzchen — APPROVED** [review]

**Review by sanketmehta28 — APPROVED** [review]

**Pull request merged**